File: java/core/src/main/java/network/aika/activations/Activation.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.activations;

import network.aika.Model;
import network.aika.Document;
import network.aika.Element;
import network.aika.ModelProvider;
import network.aika.bindingsignal.BindingSignal;
import network.aika.neurons.Synapse;
import network.aika.bindingsignal.BSType;
import network.aika.neurons.Neuron;
import network.aika.type.Obj;
import network.aika.type.relations.Relation;
import network.aika.typedefs.*;
import network.aika.fields.field.FieldOutput;
import network.aika.type.ObjImpl;
import network.aika.queue.Queue;
import network.aika.queue.QueueProvider;
import network.aika.queue.Timestamp;
import network.aika.queue.steps.Fired;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static network.aika.misc.direction.Direction.INPUT;
import static network.aika.misc.direction.Direction.OUTPUT;
import static network.aika.queue.Timestamp.NOT_SET;
import static network.aika.typedefs.ActivationDefinition.NEURON;
import static network.aika.typedefs.ActivationDefinition.SELF;

/**
 * @author Lukas Molzberger
 */
public abstract class Activation extends ObjImpl implements Element, ModelProvider, QueueProvider, Comparable<Activation> {

    public static final Comparator<Activation> ID_COMPARATOR = Comparator.comparingInt(Activation::getId);

    protected final Integer id;

    protected Neuron neuron;
    protected Document doc;

    protected Map<BSType, BindingSignal> bindingSignals;

    protected Activation parent;

    protected Timestamp created = NOT_SET;
    protected Timestamp fired = NOT_SET;
    protected Fired firedStep = new Fired(this);

    protected NavigableMap<Integer, Link> outputLinks = new TreeMap<>();

    public Activation(
            ActivationDefinition t,
            Activation parent,
            Integer id,
            Neuron n,
            Document doc,
            Map<BSType, BindingSignal> bindingSignals
    ) {
        super(t);
        this.parent = parent;
        this.id = id;
        this.neuron = n;
        this.doc = doc;
        this.bindingSignals = bindingSignals;

        doc.addActivation(this);
        neuron.updateLastUsed(doc.getId());

        setCreated(doc.getCurrentTimestamp());
    }

    @Override
    public Stream<Obj> followManyRelation(Relation rel) {
        if(rel == ActivationDefinition.INPUT)
            return getInputLinks().map(o -> o);
        else if(rel == OUTPUT)
            return getOutputLinks().map(o -> o);
        else
            throw new RuntimeException("Invalid Relation");
    }

    @Override
    public Obj followSingleRelation(Relation rel) {
        if(rel == SELF)
            return this;
        else if(rel == NEURON)
            return neuron;
        else
            throw new RuntimeException("Invalid Relation");
    }

    public ActivationKey getKey() {
        return new ActivationKey(neuron.getId(), id);
    }

    public Activation getParent() {
        return parent;
    }

    public void addOutputLink(Link l) {
        Activation oAct = l.getOutput();
        assert outputLinks.get(oAct.getId()) == null;
        outputLinks.put(oAct.getId(), l);
    }

    public abstract void addInputLink(Link l);

    public BindingSignal getBindingSignal(BSType s) {
        return bindingSignals.get(s);
    }

    public Map<BSType, BindingSignal> getBindingSignals() {
        return bindingSignals;
    }

    public boolean hasConflictingBindingSignals(Map<BSType, BindingSignal> targetBindingSignals) {
        return targetBindingSignals
                .entrySet()
                .stream()
                .anyMatch(e ->
                    isConflictingBindingSignal(e.getKey(), e.getValue())
                );
    }

    public boolean isConflictingBindingSignal(BSType s, BindingSignal targetBS) {
        BindingSignal bs = bindingSignals.get(s);
        return bs != null && targetBS != bs;
    }

    public boolean hasNewBindingSignals(Map<BSType, BindingSignal> targetBindingSignals) {
        return !bindingSignals.keySet()
                .containsAll(targetBindingSignals.keySet());
    }

    public Activation branch(Map<BSType, BindingSignal> bindingSignals) {
        // TODO: Check: Is it necessary to remove the parents binding-signals beforehand?
        bindingSignals = new HashMap<>(bindingSignals);
        getBindingSignals()
                .keySet()
                .forEach(bindingSignals::remove);

        return neuron.createActivation(
                this,
                getDocument(),
                bindingSignals
        );
    }

    public void linkOutgoing() {
        neuron.wakeupPropagable();

        neuron.getOutputSynapses()
                .stream()
                .filter(s ->
                        ((SynapseDefinition)s.getType()).isOutgoingLinkingCandidate(getBindingSignals().keySet())
                )
                .forEach(this::linkOutgoing);
    }

    void linkOutgoing(Synapse targetSyn) {
        Set<Activation> targets = collectLinkingTargets(
                targetSyn.getOutput(getModel())
        );

        targets.forEach(targetAct ->
                targetSyn.createLink(
                        this,
                        targetAct
                )
        );

        if(targets.isEmpty() && targetSyn.isPropagable())
            propagate(targetSyn);
    }

    public void propagate(Synapse targetSyn) {
        Map<BSType, BindingSignal> bindingSignals = targetSyn.transitionForward(getBindingSignals());
        Activation oAct = targetSyn.getOutput(getModel()).createActivation(null, getDocument(), bindingSignals);

        targetSyn.createLink(this, bindingSignals, oAct);

        oAct.linkIncoming(this);
    }

    public abstract void linkIncoming(Activation excludedInputAct);

    Set<Activation> collectLinkingTargets(Neuron n) {
        return getBindingSignals()
                .values()
                .stream()
                .flatMap(bs -> bs.getActivations(n))
                .collect(Collectors.toSet());
    }

    public int getId() {
        return id;
    }

    public Timestamp getCreated() {
        return created;
    }

    public void setCreated(Timestamp ts) {
        this.created = ts;
    }

    @Override
    public Timestamp getFired() {
        return fired;
    }

    public void setFired() {
        fired = doc.getCurrentTimestamp();
    }

    public void setFired(Timestamp f) {
        fired = f;
    }

    public void updateFiredStep(FieldOutput net) {
        if(!net.exceedsThreshold() || fired != NOT_SET)
            return;

        if(firedStep.isQueued())
            doc.removeStep(firedStep);

        firedStep.updateNet(net.getUpdatedValue());
        doc.addStep(firedStep);
    }

    @Override
    public Queue getQueue() {
        return doc;
    }

    public Neuron getNeuron() {
        return neuron;
    }

    public Document getDocument() {
        return doc;
    }

    @Override
    public Model getModel() {
        return getNeuron().getModel();
    }

    public Link getCorrespondingInputLink(Link l) {
        return null;
    }

    public Link getCorrespondingOutputLink(Link l) {
        return null;
    }

    public Stream<Link> getInputLinks(LinkDefinition linkDefinition) {
        return getInputLinks();
    }

    public abstract Stream<Link> getInputLinks();

    public Stream<Link> getOutputLinks(LinkDefinition linkDefinition) {
        return getOutputLinks();
    }

    public Stream<Link> getOutputLinks() {
        return outputLinks.values()
                .stream();
    }

    public Link getOutputLink(Neuron n) {
        return outputLinks.get(n.getId());
    }

    public Stream<Link> getOutputLinks(Synapse s) {
        return getOutputLinks()
                .filter(l -> l.getSynapse() == s);
    }

    @Override
    public int compareTo(Activation act) {
        return ID_COMPARATOR.compare(this, act);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Activation)) return false;
        Activation that = (Activation) o;
        return id == that.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return type.getName() + " " + toKeyString();
    }

    @Override
    public String toKeyString() {
        return "id:" + getId() + " n:[" + getNeuron().toKeyString() + "]";
    }
}

File: src/network/activation.cpp
#include "network/activation.h"
#include "network/direction.h"
#include "network/input.h"
#include "network/output.h"
#include "network/types/activation_type.h"
#include "fields/rel_obj_iterator.h"
#include "fields/field.h"
#include "network/context.h"
#include "network/fired.h"
#include "network/link.h"
#include "network/linker.h"
#include "network/binding_signal.h"
#include <cassert>
#include <stdexcept>

const std::function<bool(Activation*, Activation*)> Activation::ID_COMPARATOR = [](Activation* a1, Activation* a2) {
    return a1->getId() < a2->getId();
};

Activation::Activation(ActivationType* t, Activation* parent, int id, Neuron* n, Context* ctx, std::map<int, BindingSignal*> bindingSignals)
    : Object(t), id(id), neuron(n), ctx(ctx), bindingSignals(bindingSignals), parent(parent), created(-1), fired(-1), firedStep(new Fired(this)) {
    initFields();
    ctx->addActivation(this);
    neuron->updateLastUsed(ctx->getId());
    setCreated(ctx->getCurrentTimestamp());
}

Activation::~Activation() {
    std::cout << "~Activation" << std::endl;
    delete firedStep;
}

RelatedObjectIterable* Activation::followManyRelation(Relation* rel) const {
    // Create a custom iterable for each relation type
    if (rel->getRelationLabel() == "INPUT") {
        // Since getInputLinks() is pure virtual, derived classes should implement specialized behavior
        // This base implementation handles the common case for OUTPUT relations
        return nullptr;
    } else if (rel->getRelationLabel() == "OUTPUT") {
        // Convert getOutputLinks() vector to an iterable
        std::vector<Link*> links = const_cast<Activation*>(this)->getOutputLinks();
        std::vector<Object*> objs;
        for (Link* link : links) {
            objs.push_back(static_cast<Object*>(link));
        }
        return new VectorObjectIterable(objs);
    } else {
        throw std::runtime_error("Invalid Relation: " + rel->getRelationLabel());
    }
}

Object* Activation::followSingleRelation(const Relation* rel) const {
    if (rel->getRelationLabel() == "SELF") {
        return const_cast<Activation*>(this);
    } else if (rel->getRelationLabel() == "NEURON") {
        return neuron;
    } else {
        throw std::runtime_error("Invalid Relation");
    }
}

ActivationKey Activation::getKey() const {
    return ActivationKey(neuron->getId(), id);
}

Activation* Activation::getParent() const {
    return parent;
}

BindingSignal* Activation::getBindingSignal(int s) const {
    auto it = bindingSignals.find(s);
    if (it != bindingSignals.end()) {
        return it->second;
    }
    return nullptr;
}

std::map<int, BindingSignal*> Activation::getBindingSignals() const {
    return bindingSignals;
}

int Activation::getId() const {
    return id;
}

long Activation::getCreated() const {
    return created;
}

void Activation::setCreated(long ts) {
    created = ts;
}

long Activation::getFired() const {
    return fired;
}

void Activation::setFired() {
    fired = ctx->getCurrentTimestamp();
}

void Activation::setFired(long f) {
    fired = f;
}

void Activation::updateFiredStep(Field* net) {
    // TODO: Add exceedsThreshold method to Field class
    // if (!net->exceedsThreshold() || fired != -1) {
    if (fired != -1) {
        return;
    }

    if (firedStep->isQueued()) {
        ctx->removeStep(firedStep);
    }

    firedStep->updateNet(net->getUpdatedValue());
    ctx->addStep(firedStep);
}

Queue* Activation::getQueue() const {
    return ctx;
}

Neuron* Activation::getNeuron() const {
    return neuron;
}

Context* Activation::getContext() const {
    return ctx;
}

Model* Activation::getModel() const {
    return neuron->getModel();
}

Config* Activation::getConfig() const {
    return neuron->getModel()->getConfig();
}

std::vector<Link*> Activation::getInputLinks(LinkType* linkDefinition) const {
    return getInputLinks();
}

void Activation::addInputLink(Link* l) {
    //assert(inputLinks.find(syn->getSynapseId()) == inputLinks.end());
    inputLinks[getInputKey(l)] = l;
}

std::vector<Link*> Activation::getInputLinks() const {
    std::vector<Link*> result;
    for (const auto& pair : inputLinks) {
        result.push_back(pair.second);
    }
    return result;
}

std::vector<int> Activation::getInputKey(Link* l) const {
    std::vector<int> key;
    key.push_back(l->getSynapse()->getSynapseId());

    for (const auto& t : static_cast<SynapseType*>(l->getSynapse()->getType())->getTransitions()) {
        BindingSignal* bs = l->getInput()->getBindingSignal(t->from());
        key.push_back(bs->getTokenId());
    }

    return key;
}

void Activation::addOutputLink(Link* l) {
    //    assert(outputLinks.find(oAct->getId()) == outputLinks.end());
    outputLinks[getOutputKey(l)] = l;
}

std::vector<Link*> Activation::getOutputLinks(LinkType* linkDefinition) const {
    return getOutputLinks();
}

std::vector<Link*> Activation::getOutputLinks() const {
    std::vector<Link*> result;
    for (const auto& pair : outputLinks) {
        result.push_back(pair.second);
    }
    return result;
}

Link* Activation::getOutputLink(Neuron* n) const {
/*    auto it = outputLinks.find(n->getId());
    if (it != outputLinks.end()) {
        return it->second;
    }
*/
    return nullptr;
}

std::vector<Link*> Activation::getOutputLinks(Synapse* s) const {
    std::vector<Link*> result;
    for (auto& l : getOutputLinks()) {
        if (l->getSynapse() == s) {
            result.push_back(l);
        }
    }
    return result;
}

std::vector<int> Activation::getOutputKey(Link* l) const {
    std::vector<int> key;
    key.push_back(l->getSynapse()->getSynapseId());

    for (const auto& t : static_cast<SynapseType*>(l->getSynapse()->getType())->getTransitions()) {
        BindingSignal* bs = l->getInput()->getBindingSignal(t->from());
        key.push_back(bs->getTokenId());
    }

    return key;
}

int Activation::compareTo(Activation* act) const {
    // For const correctness, we remove const to use with ID_COMPARATOR
    return id < act->id ? -1 : (id > act->id ? 1 : 0);
}

bool Activation::equals(Activation* o) const {
    return this == o || (o != nullptr && id == o->id);
}

int Activation::hashCode() const {
    return std::hash<int>()(id);
}

std::string Activation::toString() const {
    return type->getName() + " " + toKeyString();
}

std::string Activation::toKeyString() const {
    return "id:" + std::to_string(getId()) + " n:[" + neuron->toKeyString() + "]";
}

std::vector<int> Activation::createInputKeyFromOutputCandidate(Synapse* outputSynapse, Activation* outputActivation) const {
    std::vector<int> key;
    
    // Get the paired input synapse ID from the output synapse
    int pairedInputSynapseId = outputSynapse->getPairedInputSynapseId();
    if (pairedInputSynapseId == -1) {
        // No paired input synapse - return empty key
        return key;
    }
    
    // Start building the key with the paired input synapse ID
    key.push_back(pairedInputSynapseId);
    
    // Get binding signals from the output activation
    std::map<int, BindingSignal*> outputBindingSignals = outputActivation->getBindingSignals();
    
    // Transition the binding signals backwards through the output synapse
    std::map<int, BindingSignal*> inputBindingSignals = outputSynapse->transitionBackward(outputBindingSignals);

    // Add the resolved wildcard binding signal and other binding signals to the key
    // We need to get the transitions from the paired input synapse type to know which slots to use
    SynapseType* outputSynapseType = static_cast<SynapseType*>(outputSynapse->getType());

    // Get the paired input synapse type from the pairing configs
    SynapseType* pairedInputSynapseType = nullptr;

    // Check input side pairing config
    const PairingConfig& inputConfig = outputSynapseType->getInputSidePairingConfig();
    if (inputConfig.pairedSynapseType && inputConfig.pairedSynapseType->getInstanceSynapseType()) {
        // Check if this could be our paired synapse by examining synapse IDs
        // We need a way to match the synapse ID - for now, assume it's the input side paired synapse
        pairedInputSynapseType = inputConfig.pairedSynapseType;
    }

/*
    // Check output side pairing config if we didn't find it on input side
    if (!pairedInputSynapseType) {
        const PairingConfig& outputConfig = outputSynapseType->getOutputSidePairingConfig();
        if (outputConfig.pairedSynapseType && outputConfig.pairedSynapseType->getInstanceSynapseType()) {
            pairedInputSynapseType = outputConfig.pairedSynapseType;
        }
    }
*/

    if (pairedInputSynapseType) {
        // Get transitions from the paired input synapse type
        const std::vector<Transition*> transitions = pairedInputSynapseType->getTransitions();
        
        // For each transition in the paired input synapse, add the corresponding binding signal token ID
        for (const auto& transition : transitions) {
            int fromSlot = transition->from();
            auto it = inputBindingSignals.find(fromSlot);
            if (it != inputBindingSignals.end() && it->second) {
                key.push_back(it->second->getTokenId());
            }
        }
    }
    
    return key;
}
